atom/common/platform_util_win.cc

```diff
@@ -169,33 +170,298 @@ bool OpenExternal(const GURL& url) {
+  HRESULT hrDelete,
+  __RPC__in_opt IShellItem *psiNewlyCreated) {   
+  return S_OK;
+}
+
+HRESULT DelFileProgressSink::PreNewItem( 
+  DWORD dwFlags,
+  __RPC__in_opt IShellItem *psiDestinationFolder,
+  __RPC__in_opt_string LPCWSTR pszNewName) {
+  return E_NOTIMPL;
+}
+
+HRESULT DelFileProgressSink::PostNewItem( 
+  DWORD dwFlags,
+  __RPC__in_opt IShellItem *psiDestinationFolder,
+  __RPC__in_opt_string LPCWSTR pszNewName,
+  __RPC__in_opt_string LPCWSTR pszTemplateName,
+  DWORD dwFileAttributes,
+  HRESULT hrNew,
+  __RPC__in_opt IShellItem *psiNewItem) {   
+  return E_NOTIMPL;
+}
+
+HRESULT DelFileProgressSink::UpdateProgress( 
+  UINT iWorkTotal,
+  UINT iWorkSoFar) {   
+  return S_OK;
+}
+
+HRESULT DelFileProgressSink::ResetTimer() {   
+  return S_OK;
+}
+
+HRESULT DelFileProgressSink::PauseTimer() {   
+  return S_OK;
+}
+
+HRESULT DelFileProgressSink::ResumeTimer() {   
+  return S_OK;
+}
+
 bool MoveItemToTrash(const base::FilePath& path) {
-  // SHFILEOPSTRUCT wants the path to be terminated with two NULLs,
-  // so we have to use wcscpy because wcscpy_s writes non-NULLs
-  // into the rest of the buffer.
-  wchar_t double_terminated_path[MAX_PATH + 1] = {0};
-#pragma warning(suppress:4996)  // don't complain about wcscpy deprecation
-  wcscpy(double_terminated_path, path.value().c_str());
-
-  SHFILEOPSTRUCT file_operation = {0};
-  file_operation.wFunc = FO_DELETE;
-  file_operation.pFrom = double_terminated_path;
-  file_operation.fFlags = FOF_ALLOWUNDO | FOF_SILENT | FOF_NOCONFIRMATION;
-  int err = SHFileOperation(&file_operation);
-
-  // Since we're passing flags to the operation telling it to be silent,
-  // it's possible for the operation to be aborted/cancelled without err
-  // being set (although MSDN doesn't give any scenarios for how this can
-  // happen).  See MSDN for SHFileOperation and SHFILEOPTSTRUCT.
-  if (file_operation.fAnyOperationsAborted)
-    return false;
+  bool retResult = false;
+  
+  HRESULT hr = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE); 
+  if (SUCCEEDED(hr)) {
+    IFileOperation *pfo;
+    hr = CoCreateInstance(CLSID_FileOperation, NULL, CLSCTX_ALL, IID_PPV_ARGS(&pfo));
+    if (SUCCEEDED(hr)) {
+      // Elevation prompt enabled for UAC protected files.  This overrides the SILENT,
+      // NO_UI and NOERRORUI flags.
+      hr = pfo->SetOperationFlags(FOF_NO_UI |
+                                  FOF_ALLOWUNDO |
+                                  FOF_NOERRORUI |
+                                  FOF_SILENT |
+                                  FOFX_SHOWELEVATIONPROMPT |
+                                  FOFX_RECYCLEONDELETE);
+      if (SUCCEEDED(hr)) {
+        // Create an IShellItem from the supplied source path.
+        IShellItem *psiDel = NULL;
+        hr = SHCreateItemFromParsingName(path.value().c_str(), NULL, IID_IShellItem, (void**)&psiDel);
+        if (SUCCEEDED(hr)) {
+            DelFileProgressSink* del_sink = new DelFileProgressSink();
```

Comment:
This leaks, you heap allocate and never call delete. This should be stack allocated

DeleteFileProgresssing del_sink;

then all -> turns into ., but no heap allocation and automatic clean-up when the stack unwinds (however that happens, normal return, early return, exception, etc..)